cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import com.google.common.base.Strings;
+import com.google.common.base.Supplier;
+import com.google.common.util.concurrent.AbstractIdleService;
+import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfigValidationException;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.LsRemoteCommand;
+import org.eclipse.jgit.api.TransportCommand;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.api.errors.TransportException;
+import org.eclipse.jgit.internal.storage.file.FileRepository;
+import org.eclipse.jgit.lib.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import javax.annotation.Nullable;
+
+/**
+ * A git repository manager that is responsible for handling interfacing with git. It fetches credentials
+ * from a {@link io.cdap.cdap.api.security.store.SecureStore}, builds an {@link AuthenticationStrategy} for different
+ * Git hosting services and provides version control operations.
+ */
+public class RepositoryManager extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(RepositoryManager.class);
+  private final SecureStore secureStore;
+  private final SourceControlConfig sourceControlConfig;
+  private Git git;
+
+  public RepositoryManager(SecureStore secureStore, SourceControlConfig sourceControlConfig) {
+    this.secureStore = secureStore;
+    this.sourceControlConfig = sourceControlConfig;
+  }
+
+  /**
+   * Returns the base path in the git repo to store CDAP applications. If an optional Path prefix is provided in the
+   * repository configuration, the returned path includes it.
+   *
+   * @return the path for the repository base directory.
+   */
+  public Path getBasePath() {
```

Comment:
Is this an unused method? I don't see any usage of it?

Reply:
I think this will be used by the caller, who needs to get the path to write files/makes changes to and then push?

Reply:
Like using git from command line, the repository manager will only be called when VCS operation like clone, push, pull, etc are required. The users are expected to interact with the filesystem directly. This method returns the directory in the filesystem to be used for storing CDAP applications, i.e where the repo is cloned.

Reply:
I checked out the code and there is reference to this method. Am I missing something?

Reply:
The method will be used by Classes created by @samdgupi and @de-lan in the SourceControl operation runners. To know in which directory to add files before committing, the operation runner needs to know the directory being tracked by Git. Since Repo manager can create subdirectories in the directory configured by cdap-site.xml and repoConfig can contain a path relative to repository root to store CDAP applications, the RepositoryManager needs to queried to the path.